File: services/api-phoenix/lib/app_web/controllers/auth_controller.ex
defmodule AppWeb.AuthController do
  use Phoenix.Controller, formats: [:json]

  import Plug.Conn
  alias App.Accounts
  alias App.Auth.Token
  alias AppWeb.Plugs.RequireAuth

  @github_auth_url "https://github.com/login/oauth/authorize"
  @github_token_url "https://github.com/login/oauth/access_token"
  @github_user_url "https://api.github.com/user"

  # GET /auth/login — redirect to GitHub OAuth
  def login(conn, _params) do
    client_id = System.get_env("GITHUB_CLIENT_ID") || ""
    redirect_uri = System.get_env("GITHUB_CALLBACK_URL") || ""

    url =
      "#{@github_auth_url}?client_id=#{client_id}" <>
        "&redirect_uri=#{URI.encode_www_form(redirect_uri)}" <>
        "&scope=user:email"

    conn
    |> put_resp_header("location", url)
    |> send_resp(302, "")
  end

  # GET /auth/callback — exchange code, upsert user, return JWT
  def callback(conn, %{"code" => code}) do
    with {:ok, access_token} <- exchange_code(code),
         {:ok, gh_user} <- fetch_github_user(access_token),
         {:ok, user} <- upsert_from_github(gh_user) do
      token = Token.sign_token(user.id, user.email, user.name)
      json(conn, %{token: token, user: user_json(user)})
    else
      {:error, reason} ->
        conn
        |> put_status(400)
        |> json(%{error: "OAuth failed", reason: inspect(reason)})
    end
  end

  def callback(conn, _params) do
    conn
    |> put_status(400)
    |> json(%{error: "Missing OAuth code"})
  end

  # GET /auth/me — return current user (requires auth)
  def me(conn, _params) do
    conn = RequireAuth.call(conn, [])

    if conn.halted do
      conn
    else
      json(conn, %{user: user_json(conn.assigns.current_user)})
    end
  end

  # POST /auth/logout — stateless JWT; client discards token
  def logout(conn, _params) do
    json(conn, %{message: "Logged out"})
  end

  # POST /auth/api-key — regenerate API key (requires auth)
  def create_api_key(conn, _params) do
    conn = RequireAuth.call(conn, [])

    if conn.halted do
      conn
    else
      {:ok, user} = Accounts.regenerate_api_key(conn.assigns.current_user)
      json(conn, %{api_key: user.api_key})
    end
  end

  # DELETE /auth/api-key — revoke API key (requires auth)
  def delete_api_key(conn, _params) do
    conn = RequireAuth.call(conn, [])

    if conn.halted do
      conn
    else
      {:ok, _user} = Accounts.clear_api_key(conn.assigns.current_user)
      json(conn, %{message: "API key revoked"})
    end
  end

  # POST /dev/token — development-only token (guarded by Mix.env())
  def dev_token(conn, params) do
    if Mix.env() == :dev do
      user_id = params["user_id"] || "dev-user-id"
      email = params["email"] || "dev@example.com"
      name = params["name"] || "Dev User"
      token = Token.sign_token(user_id, email, name)
      json(conn, %{token: token})
    else
      conn
      |> put_status(404)
      |> json(%{error: "Not found"})
    end
  end

  # --- Private helpers ---

  defp exchange_code(code) do
    client_id = System.get_env("GITHUB_CLIENT_ID") || ""
    client_secret = System.get_env("GITHUB_CLIENT_SECRET") || ""

    body =
      URI.encode_query(%{
        client_id: client_id,
        client_secret: client_secret,
        code: code
      })

    case :httpc.request(
           :post,
           {String.to_charlist(@github_token_url), [{'accept', 'application/json'}],
            'application/x-www-form-urlencoded', body},
           [],
           []
         ) do
      {:ok, {{_, 200, _}, _headers, resp_body}} ->
        decoded = Jason.decode!(to_string(resp_body))
        {:ok, decoded["access_token"]}

      _ ->
        {:error, :github_token_exchange_failed}
    end
  end

  defp fetch_github_user(access_token) do
    case :httpc.request(
           :get,
           {String.to_charlist(@github_user_url),
            [
              {'authorization', String.to_charlist("Bearer #{access_token}")},
              {'user-agent', 'phoenix-app'}
            ]},
           [],
           []
         ) do
      {:ok, {{_, 200, _}, _headers, resp_body}} ->
        {:ok, Jason.decode!(to_string(resp_body))}

      _ ->
        {:error, :github_user_fetch_failed}
    end
  end

  defp upsert_from_github(gh_user) do
    Accounts.upsert_user(%{
      email: gh_user["email"] || "#{gh_user["login"]}@github.invalid",
      name: gh_user["name"] || gh_user["login"],
      provider: "github"
    })
  end

  defp user_json(user) do
    %{
      id: user.id,
      email: user.email,
      name: user.name,
      provider: user.provider
    }
  end
end

File: services/api-rails/app/controllers/auth_controller.rb
require_relative '../../lib/jwt_service'

class AuthController < ActionController::API
  include Authenticatable

  GITHUB_AUTH_URL = 'https://github.com/login/oauth/authorize'.freeze
  GITHUB_TOKEN_URL = 'https://github.com/login/oauth/access_token'.freeze
  GITHUB_USER_URL = 'https://api.github.com/user'.freeze

  # Skip global require_auth for unauthenticated routes.
  skip_before_action :require_auth, only: %i[login callback dev_token]

  # GET /auth/login — redirect to GitHub OAuth
  def login
    params_str = URI.encode_www_form(
      client_id: ENV['GITHUB_CLIENT_ID'],
      redirect_uri: ENV['GITHUB_CALLBACK_URL'],
      scope: 'user:email'
    )
    redirect_to "#{GITHUB_AUTH_URL}?#{params_str}", allow_other_host: true
  end

  # GET /auth/callback — exchange code, upsert user, return JWT
  def callback
    code = params[:code]
    return render json: { error: 'Missing OAuth code' }, status: :bad_request unless code

    access_token = exchange_code(code)
    return render json: { error: 'GitHub token exchange failed' }, status: :bad_gateway unless access_token

    gh_user = fetch_github_user(access_token)
    return render json: { error: 'GitHub user fetch failed' }, status: :bad_gateway unless gh_user

    user = upsert_from_github(gh_user)
    token = JwtService.encode({ sub: user.id, email: user.email, name: user.name })
    render json: { token: token, user: user_json(user) }
  end

  # GET /auth/me — return current user (require_auth runs via before_action)
  def me
    render json: { user: user_json(@current_user) }
  end

  # POST /auth/logout — stateless JWT; client discards token
  def logout
    render json: { message: 'Logged out' }
  end

  # POST /auth/api-key — regenerate API key
  def create_api_key
    @current_user.regenerate_api_key!
    render json: { api_key: @current_user.api_key }
  end

  # DELETE /auth/api-key — revoke API key
  def delete_api_key
    @current_user.update!(api_key: nil)
    render json: { message: 'API key revoked' }
  end

  # POST /dev/token — development-only token
  def dev_token
    unless Rails.env.development?
      return render json: { error: 'Not found' }, status: :not_found
    end

    user_id = params[:user_id] || 'dev-user-id'
    email = params[:email] || 'dev@example.com'
    name = params[:name] || 'Dev User'
    token = JwtService.encode({ sub: user_id, email: email, name: name })
    render json: { token: token }
  end

  private

  def exchange_code(code)
    uri = URI(GITHUB_TOKEN_URL)
    response = Net::HTTP.post_form(uri, {
      client_id: ENV['GITHUB_CLIENT_ID'],
      client_secret: ENV['GITHUB_CLIENT_SECRET'],
      code: code
    })
    body = JSON.parse(response.body) rescue {}
    body['access_token']
  end

  def fetch_github_user(access_token)
    uri = URI(GITHUB_USER_URL)
    req = Net::HTTP::Get.new(uri)
    req['Authorization'] = "Bearer #{access_token}"
    req['Accept'] = 'application/json'
    req['User-Agent'] = 'rails-app'

    res = Net::HTTP.start(uri.hostname, uri.port, use_ssl: true) { |http| http.request(req) }
    JSON.parse(res.body) rescue nil
  end

  def upsert_from_github(gh_user)
    email = gh_user['email'] || "#{gh_user['login']}@github.invalid"
    user = User.find_or_initialize_by(email: email)
    user.assign_attributes(
      name: gh_user['name'] || gh_user['login'],
      provider: 'github'
    )
    user.save!
    user
  end

  def user_json(user)
    { id: user.id, email: user.email, name: user.name, provider: user.provider }
  end
end
